Address Safer CPP failures in JSContext

cdumez opened the pull request:
#### 4903b29014c50616d45cb769528d8cfe6938d3bd
<pre>
Address Safer CPP failures in JSContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=288409">https://bugs.webkit.org/show_bug.cgi?id=288409</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/API/JSContext.mm:
(+[JSContext currentContext]):
(+[JSContext currentThis]):
(+[JSContext currentCallee]):
(+[JSContext currentArguments]):
(-[JSContext name]):
(-[JSContext beginCallbackWithData:calleeValue:thisValue:argumentCount:arguments:]):
(-[JSContext endCallbackWithData:]):
* Source/JavaScriptCore/API/JSContextRef.cpp:
(JSContextGroupSetExecutionTimeLimit):
(JSContextGroupClearExecutionTimeLimit):
(JSGlobalContextSetDebuggerRunLoop):
* Source/JavaScriptCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/JavaScriptCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::protectedInspectorDebuggable):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:

Canonical link: <a href="https://commits.webkit.org/291186@main">https://commits.webkit.org/291186@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3b125ca14b19077c517dd0a0bcef44982acc35f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92259 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1353 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/42795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94309 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/12103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20268 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/42795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95260 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/12103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/12103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/1163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/84998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/12103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/90949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/19336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/99296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/19588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/79410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19571 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/19318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113575 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/19010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/20751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -208,24 +208,24 @@ - (JSValue *)globalObject
 {
-    Thread& thread = Thread::current();
-    CallbackData *entry = (CallbackData *)thread.m_apiData;
+    Ref thread = Thread::current();
```
This looks strange as it is thread local data tied to the thread itself.

cdumez replied:
@geoffreygaren @rniwa : Should we rename `Thread::current()` to `Thread::currentSingleton()`? Would that make the static analyzer happy? Or should I just suppress all those warnings?

geoffreygaren replied:
currentSingleton seems close to ok — but does that only ignore local variables or also ignore lambda capture and/or stores to the heap?

Can we put SUPPRESS_UNCOUNTED_LOCAL on the function’s return value, or does it have to be on the local variable?

rniwa replied:
It has to be on the local variable declaration.

geoffreygaren replied:
@rniwa Does ‘Singleton’ suppress all warnings, or just local variable warnings?

rniwa replied:
It affects call arguments & local variables.

Constellation commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -208,24 +208,24 @@ - (JSValue *)globalObject
-    Thread& thread = Thread::current();
-    CallbackData *entry = (CallbackData *)thread.m_apiData;
+    Ref thread = Thread::current();
+    CallbackData *entry = (CallbackData *)thread->m_apiData;
```
Ditto

Constellation commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -208,24 +208,24 @@ - (JSValue *)globalObject
 {
-    Thread& thread = Thread::current();
-    CallbackData *entry = (CallbackData *)thread.m_apiData;
+    Ref thread = Thread::current();
```
Ditto

Constellation commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -234,8 +234,8 @@ + (JSValue *)currentCallee
-    Thread& thread = Thread::current();
-    CallbackData *entry = (CallbackData *)thread.m_apiData;
+    Ref thread = Thread::current();
+    CallbackData *entry = (CallbackData *)thread->m_apiData;
```
Ditto

Constellation commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -379,19 +379,19 @@ - (BOOL)boolFromNotifyException:(JSValueRef)exceptionValue
 - (void)beginCallbackWithData:(CallbackData *)callbackData calleeValue:(JSValueRef)calleeValue thisValue:(JSValueRef)thisValue argumentCount:(size_t)argumentCount arguments:(const JSValueRef *)arguments
 {
-    Thread& thread = Thread::current();
+    Ref thread = Thread::current();
```
Ditto

Constellation commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -379,19 +379,19 @@ - (BOOL)boolFromNotifyException:(JSValueRef)exceptionValue
 - (void)endCallbackWithData:(CallbackData *)callbackData
 {
-    Thread& thread = Thread::current();
+    Ref thread = Thread::current();
```
Ditto

darinadler commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -259,11 +259,11 @@ - (JSVirtualMachine *)virtualMachine
 - (NSString *)name
 {
-    auto name = adopt(JSGlobalContextCopyName(m_context));
+    SUPPRESS_UNCOUNTED_ARG auto name = adopt(JSGlobalContextCopyName(m_context));
```
I don't understand this one. Since `name` is a smart pointer I wouldn't expect a diagnostic here. Maybe the issue is with the argument to the `adopt` function?

cdumez replied:
I believe this is a false positive (cc @rniwa). It is about the argument to `adopt()`.

darinadler commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -259,11 +259,11 @@ - (JSVirtualMachine *)virtualMachine
         return nil;
 
-    return adoptCF(JSStringCopyCFString(kCFAllocatorDefault, name.get())).bridgingAutorelease();
+    SUPPRESS_UNCOUNTED_ARG return adoptCF(JSStringCopyCFString(kCFAllocatorDefault, name.get())).bridgingAutorelease();
```
I also don't fully understand this. Is this working around a deficiency in our checker?

cdumez replied:
cc @rniwa. This is a false positive. The static analysis is complaining about `name` being uncounted and unsafe.

cdumez replied:
I filed rdar://145661220 and will add a FIXME comment.

cdumez committed:
Address Safer CPP failures in JSContext
https://bugs.webkit.org/show_bug.cgi?id=288409

Reviewed by Darin Adler.

* Source/JavaScriptCore/API/JSContext.mm:
(+[JSContext currentContext]):
(+[JSContext currentThis]):
(+[JSContext currentCallee]):
(+[JSContext currentArguments]):
(-[JSContext name]):
(-[JSContext beginCallbackWithData:calleeValue:thisValue:argumentCount:arguments:]):
(-[JSContext endCallbackWithData:]):
* Source/JavaScriptCore/API/JSContextRef.cpp:
(JSContextGroupSetExecutionTimeLimit):
(JSContextGroupClearExecutionTimeLimit):
(JSGlobalContextSetDebuggerRunLoop):
* Source/JavaScriptCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/JavaScriptCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::protectedInspectorDebuggable):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:

Canonical link: https://commits.webkit.org/291186@main
